### PR TITLE
Merge upstream changes, Fix many issues

### PR DESCRIPTION
**Pull request by Ralender**
in this patch:
 - merge upstream changes
 - improve documentation
 - fix remaining sw_emu and hw_emu tests
 - fix all but one hw tests
 - improve usability of the compiler
 - enable lowering clang annotation into IR compatible HLS pragmas

### CODE REVIEW

File: llvm/lib/SYCL/LowerSYCLMetaData.cpp [hunk 1]
```diff
@@ -0,0 +1,176 @@
+//===- LowerSYCLMetaData.cpp ---------------===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+//
+// Lower optimization metadata. This Pass must be run before inling.
```
Comment (keryell) [inline llvm/lib/SYCL/LowerSYCLMetaData.cpp]: Ellaborate

File: llvm/lib/SYCL/LowerSYCLMetaData.cpp [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+#include <string>
+
+#include "llvm/Analysis/LoopInfo.h"
+#include "llvm/Analysis/ValueTracking.h"
+#include "llvm/IR/Dominators.h"
+#include "llvm/IR/CallingConv.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/SYCL/LowerSYCLMetaData.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/ADT/DenseMap.h"
```
Comment (keryell) [inline llvm/lib/SYCL/LowerSYCLMetaData.cpp]: To sort?

File: llvm/lib/SYCL/LowerSYCLMetaData.cpp [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+          if (L)
+            return Functor(L);
+          if (auto* I = dyn_cast<Instruction>(Usr))
+            Stack.insert(I->getFunction());
+        };
+        if (auto *CB = dyn_cast<CallBase>(Usr))
+          if (U.getOperandNo() == CB->arg_size())
+            OnLoop(findLoopAround(CB));
+        
+        //TODO: Try follow uses to deal with function pointers.
```
Comment (keryell) [inline llvm/lib/SYCL/LowerSYCLMetaData.cpp]: Syntax?

File: llvm/lib/SYCL/LowerSYCLMetaData.cpp [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+  LSMDState(Module& M) : M(M), Ctx(M.getContext()) {}
+  Module& M;
+  LLVMContext& Ctx;
+  bool HasChanged = false;
+
+  llvm::SmallDenseMap<Function*, std::unique_ptr<DominatorTree>, 16> DTCache;
+  llvm::SmallDenseMap<Function*, std::unique_ptr<LoopInfo>, 16> LICache;
+  // llvm::SmallDenseSet<Function*, 8>
+
+  Loop* findLoopAround(Instruction* I) {
```
Comment (keryell) [inline llvm/lib/SYCL/LowerSYCLMetaData.cpp]: Add comments.
No write-only code...

File: llvm/lib/SYCL/LowerSYCLMetaData.cpp [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+    if (!LI) {
+      std::unique_ptr<DominatorTree> &DT = DTCache[F];
+      if (!DT)
+        DT = std::make_unique<DominatorTree>(*F);
+      LI = std::make_unique<LoopInfo>(*DT);
+    }
+    return LI->getLoopFor(I->getParent());
+  }
+
+  void findLoopAroundFunction(Function *Start, function_ref<void(Loop *)> Functor) {
```
Comment (keryell) [inline llvm/lib/SYCL/LowerSYCLMetaData.cpp]: Add some prose along the code... Or poetry, haiku, alexandrins if you prefer...

File: sycl/include/CL/sycl/xilinx/fpga/partition_array.hpp [hunk 6]
```diff
@@ -39,13 +39,17 @@ namespace partition {
...
+  enum type : int {
     cyclic,
     block,
     complete,
     none
   };
+  }
 
+  __attribute__((annotate("xilinx_partition_array"))) __attribute__((always_inline))
+  inline void xilinx_partition_array(void*, int, int, int) {}
```
Comment (keryell) [inline sycl/include/CL/sycl/xilinx/fpga/partition_array.hpp]: I wonder whether it is really better to look for some annotated functions with a specific use case compared to looking for a function with a specific (mangled) name with a specific use case.

**Review by keryell — COMMENTED** [review]
Globally this looks good.
There are probably room for some improvement later in an another PR, such as improving the level of comments in the code...

File: sycl/include/CL/sycl/xilinx/fpga/partition_array.hpp (hunk 6 above)
Comment (Ralender) [reply sycl/include/CL/sycl/xilinx/fpga/partition_array.hpp]: I think there is good reasons to tag the function and not just read the mangling:
 - when refactoring prevent the mangling from becoming out of date.
 - easily allows the function to be templated

**Pull request merged**